peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package abstractlist
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/introspection"
+	intyarpcerrors "go.uber.org/yarpc/internal/yarpcerrors"
+	"go.uber.org/yarpc/pkg/lifecycle"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+var (
+	_noContextDeadlineError = "%q peer list can't wait for peer without a context deadline"
+	_unavailableError       = "%q peer list timed out waiting for peer: %s"
+)
+
+// Implementation is a collection of available peers, with its own
+// subscribers for peer status change notifications.
+// The available peer list encapsulates the logic for selecting from among
+// available peers, whereas a ChooserList is responsible for retaining,
+// releasing, and monitoring peer availability.
+// Use "go.uber.org/yarpc/peer/abstractlist".List in conjunction with an
+// Implementation to produce a "go.uber.org/yarpc/api/peer".List.
+//
+// abstractlist.List and abstractlist.Implementation compose well with sharding
+// schemes the degenerate to returning the only available peer.
+//
+// The abstractlist.List calls Add, Remove, and Choose under a write lock so
+// the implementation is free to perform mutations on its own data without
+// locks.
+type Implementation interface {
+	Add(peer.StatusPeer, peer.Identifier) Subscriber
+	Remove(peer.StatusPeer, peer.Identifier, Subscriber)
+	// Choose must return an available peer under a list read lock, so must
+	// not block.
+	Choose(context.Context, *transport.Request) peer.StatusPeer
+}
+
+// Subscriber is a callback that implementations of peer list data structures
+// must provide.
+//
+// The peer list uses the Subscriber to send notifications when a peer’s
+// pending request count changes.
+// A peer list implementation may have a single subscriber or a subscriber for
+// each peer.
+//
+// Peer list data structure implementations
+type Subscriber interface {
+	UpdatePendingRequestCount(peer.Identifier, int)
```

Comment:
Implementations are free to reuse a single subscriber for every peer. None of them do anymore, though. It would be reasonable to remove the additional argument at this stage.

Reply:
If none of them do, it might make sense to remove it. Otherwise, let's add a comment that this is the original identifier used when the peer was added